compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fold_dma_waits.mlir

```diff
@@ -220,3 +220,93 @@ module attributes {hal.executable.target = #executable_target_amdaie_xclbin_fb}
+      %8 = amdaie.logicalobjectfifo.placeholder{%tile_3_0} : !amdaie.logicalobjectfifo<memref<64x32xi32>>
+      %channel = amdaie.channel(%tile_0_0, 0, port_type = DMA, direction = MM2S)
+      %channel_10 = amdaie.channel(%tile_0_1, 0, port_type = DMA, direction = S2MM)
+      %channel_11 = amdaie.channel(%tile_1_0, 0, port_type = DMA, direction = MM2S)
+      %channel_12 = amdaie.channel(%tile_1_1, 0, port_type = DMA, direction = S2MM)
+      %channel_13 = amdaie.channel(%tile_3_0, 0, port_type = DMA, direction = MM2S)
+      %channel_14 = amdaie.channel(%tile_3_1, 0, port_type = DMA, direction = S2MM)
+      %9 = amdaie.flow({%channel} -> {%channel_10}) {is_packet_flow = false}
+      %10 = amdaie.connection(%0 {%channel_10}, %2 {%channel}, flow = %9) {connection_type = #amdaie<connection_type Packet>} : (!amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>, !amdaie.logicalobjectfifo<memref<64x32xi32>>)
+      %11 = amdaie.flow({%channel_11} -> {%channel_12}) {is_packet_flow = false}
+      %12 = amdaie.connection(%3 {%channel_12}, %5 {%channel_11}, flow = %11) {connection_type = #amdaie<connection_type Packet>} : (!amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>, !amdaie.logicalobjectfifo<memref<64x32xi32>>)
+      %13 = amdaie.flow({%channel_13} -> {%channel_14}) {is_packet_flow = false}
+      %14 = amdaie.connection(%6 {%channel_14}, %8 {%channel_13}, flow = %13) {connection_type = #amdaie<connection_type Packet>} : (!amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>, !amdaie.logicalobjectfifo<memref<64x32xi32>>)
+      amdaie.controlcode {
+        %15 = amdaie.logicalobjectfifo.from_memref %1, {%tile_0_0} : memref<64x32xi32> -> !amdaie.logicalobjectfifo<memref<2048xi32>>
+        memref.assume_alignment %1, 64 : memref<64x32xi32>
+        %16 = amdaie.logicalobjectfifo.from_memref %4, {%tile_1_0} : memref<64x32xi32> -> !amdaie.logicalobjectfifo<memref<2048xi32>>
+        memref.assume_alignment %4, 64 : memref<64x32xi32>
+        %17 = amdaie.logicalobjectfifo.from_memref %7, {%tile_3_0} : memref<64x32xi32> -> !amdaie.logicalobjectfifo<memref<2048xi32>>
+        memref.assume_alignment %7, 64 : memref<64x32xi32>
+        %bd_id = amdaie.bd_id(%tile_0_0, %c0)
+        %18 = amdaie.npu.half_dma_cpy_nd async %10(%15 [] [] [] bd_id = %bd_id channel = %channel) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+        %bd_id_15 = amdaie.bd_id(%tile_1_0, %c0)
+        %19 = amdaie.npu.half_dma_cpy_nd async %12(%16 [] [] [] bd_id = %bd_id_15 channel = %channel_11) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+        %bd_id_16 = amdaie.bd_id(%tile_3_0, %c0)
+        %20 = amdaie.npu.half_dma_cpy_nd async %14(%17 [] [] [] bd_id = %bd_id_16 channel = %channel_13) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+        amdaie.npu.dma_wait(%18 : !amdaie.async_token)
+        amdaie.npu.dma_wait(%19 : !amdaie.async_token)
+        amdaie.npu.dma_wait(%20 : !amdaie.async_token)
```

Comment:
Could you add tests for more complex cases, cases that should fail and potential corner cases? For example:

Assuming `%connection_0` and `%connection_1` use the same channel, direction, row:
```
%0 = amdaie.npu.half_dma_cpy_nd %connection_0 
amdaie.npu.dma_wait(%0 : !amdaie.async_token)
%1 = amdaie.npu.half_dma_cpy_nd %connection_0 
%2 = amdaie.npu.half_dma_cpy_nd %connection_1
amdaie.npu.dma_wait(%2 : !amdaie.async_token)
```
In this case, the `%0` and `%2` tokens should not be batched as `%1 = amdaie.npu.half_dma_cpy_nd %connection_0` touches the same connection in between.